docs/dev/core/internal/upload.md

```diff
@@ -7,6 +7,62 @@ The Glean SDK leverages the available platform capabilities to implement any net
 Glean core controls all upload and coordinates the platform side with its own internals.
 All language bindings implement ping uploading around a common API and protocol.
 
+## The upload module in the language bindings
+This paragraph illustrates the recommended structure of the upload module implementation in the language bindings (regardless of the target language), in order to allow greater flexibility with respect to potential future requirements and features.
+
+```mermaid
+classDiagram
+    class UploadResult {
+        ~ToFFI()* int
+    }
+
+    class HttpResponse {
+        int statusCode
+        ~ToFFI() int
+    }
+
+    class UnrecoverableFailure {
+        ~ToFFI() int
+    }
+
+    class RecoverableFailure {
+        ~ToFFI() int
+    }
+
+    class PingUploader {
+        <<interface>>
+        +Upload() UploadResult
+    }
+
+    class BaseUploader {
+        +BaseUploader(PingUploader)
+        ~TriggerUploads()
+        ~CancelUploads()
+    }
+
+    class HttpUploader {
+        +Upload() UploadResult
+    }
+
+    UploadResult <|-- HttpResponse
+    UploadResult <|-- UnrecoverableFailure
+    UploadResult <|-- RecoverableFailure
+    PingUploader <|-- HttpUploader
+    PingUploader o-- BaseUploader
+
+```
+
+* The `PingUploader` interface describes the contract between the `BaseUploader` and the SDK or user-provided upload modules.
+* There is a `BaseUploader` component that is responsible for interfacing with the lower level FFI `get_upload_*` calls and dealing with the logic in a platform-coherent way.
```

Comment:
```suggestion
* There is a `BaseUploader` component that is responsible for interfacing with the lower level FFI `get_upload_task` calls and dealing with the logic in a platform-coherent way.
```

There is just the `get_upload_task` function and no other "getter"s. No need for asterisk I think.

Reply:
Also, not all language bindings use FFI calls, so I think we can skip the word `FFI` here.